mne_connectivity/var.py

```diff
@@ -0,0 +1,406 @@
+        'compute_fb_operator': compute_fb_operator
+    }
+
+    # compute the A matrix for all Epochs
+    A_mats = np.zeros((n_epochs, n_nodes * model_order, n_nodes))
+    if n_jobs == 1:
+        for idx in tqdm(range(n_epochs)):
+            A = _compute_lds_func(data[idx, ...], **model_params)
+            A_mats[idx, ...] = A
+    else:
+        try:
+            from joblib import Parallel, delayed, dump, load
+        except ImportError as e:
+            raise ImportError(e)
+
+        if memmap:
+            folder = Path("./joblib_memmap")
+            folder.mkdir(exist_ok=True, parents=True)
+
+            # dump data into memmap for joblib
+            data_filename_memmap = os.path.join(folder, "data_memmap")
+            dump(data, data_filename_memmap)
+            arr = load(data_filename_memmap, mmap_mode="r")
+        else:
+            arr = data
+
+        # run parallelized job to compute fragility over all windows
+        results = Parallel(n_jobs=n_jobs)(
+            delayed(_compute_lds_func)(
+                arr[idx, ...], **model_params
+            )
+            for idx in tqdm(range(n_epochs))
+        )
+        for idx in range(len(results)):
+            adjmat = results[idx]
+            # add additional order models in dynamic connectivity
+            # along the first node axes
+            for jdx in range(model_order):
+                A_mats[idx, jdx * n_nodes: n_nodes * (jdx + 1), :] = adjmat[
+                    -n_nodes:, jdx * n_nodes: n_nodes * (jdx + 1)
+                ]
+
+        if memmap:
+            try:
+                shutil.rmtree(folder)
+            except:  # noqa
+                print("Could not clean-up joblib memmap automatically.")
+
+    # create connectivity
+    A_mats = A_mats.reshape((n_epochs, -1))
+    conn = EpochConnectivity(data=A_mats, n_nodes=n_nodes, names=names,
+                             n_epochs_used=n_epochs,
+                             times_used=times,
+                             method='Time-varying LDS',
+                             **model_params)
+    return conn
+
+
+def _compute_lds_func(data, model_order, delta, compute_fb_operator,
+                      random_state):
+    """Compute linear system using VAR model.
+
+    Allows for parallelization over epochs.
+    """
+    n_times = data.shape[-1]
+
+    # create large snapshot with time-lags of order specified by
+    # ``order`` value
+    snaps = _construct_snapshots(
+        data, order=model_order, n_times=n_times
+    )
+
+    # get the time-shifted components of each
+    X, Y = snaps[:, :-1], snaps[:, 1:]
+
+    # use scikit-learn Ridge Regression to fit
+    fit_intercept = False
+    normalize = False
+    solver = 'auto'
+    clf = Ridge(
+        alpha=delta,
+        fit_intercept=fit_intercept,
+        normalize=normalize,
+        solver=solver,
+        random_state=random_state,
+    )
+
+    # n_samples X n_features and n_samples X n_targets
+    clf.fit(X.T, Y.T)
```

Comment:
I suspect that using sklearn here is very memory inefficient. To compute a Ridge you need to assemble X.T @ X and  X.T @ y and then solve a linear system. Time series packages assemble X.T @ X and X.T @ y on the fly by looping of the signals time stamps. This avoids a huge matrix allocation for storing X (data with all delays etc.)

Reply:
Not sure I 100% follow. Do you mean something as in statsmodels?

I didn't see this in statsmodels https://www.statsmodels.org/stable/_modules/statsmodels/tsa/vector_ar/var_model.html#VAR